src/ParameterDistributions.jl

```diff
@@ -0,0 +1,321 @@
+    return size(d.distribution_samples)[2]
+end
+
+"""
+    struct ParameterDistribution
+
+Structure to hold a parameter distribution
+"""
+struct ParameterDistribution{PDType <: ParameterDistributionType, CType <: ConstraintType}
+    distribution::PDType
+    constraints::Array{CType}
+    name::String
+
+    ParameterDistribution(parameter_distribution::PDType,
+                          constraints::Array{CType},
+                          name::String) where {PDType <: ParameterDistributionType,
+                                               CType <: ConstraintType} = !(dimension(parameter_distribution) ==
+len(constraints)) ? throw(DimensionMismatch("There must be one constraint per parameter")) : new{PDType,CType}(parameter_distribution, constraints, name)     
+end
+    
+
+"""
+    struct ParameterDistributions
+
+Structure to hold an array of ParameterDistribution's
+"""
+struct ParameterDistributions
+    parameter_distributions::Array{ParameterDistribution}
+end
+
+
+## Functions
+
+"""
+    get_name(pds::ParameterDistributions)
+
+Returns a list of ParameterDistribution names
+"""
+function get_name(pds::ParameterDistributions)
+    return [get_name(pd) for pd in pds.parameter_distributions]
+end
+
+function get_name(pd::ParameterDistribution)
+    return pd.name
+end
+
+"""
+    get_distributions(pds::ParameterDistributions)
+
+Returns a `Dict` of `ParameterDistribution` distributions by name, (unless sample type)
+"""
+function get_distribution(pds::ParameterDistributions)
+    return Dict{String,Any}(get_name(pd) => get_distribution(pd) for pd in pds.parameter_distributions)
+end
+
+function get_distribution(pd::ParameterDistribution)
+    return get_distribution(pd.distribution)
+end
+
+function get_distribution(d::Samples)
+    return "Contains samples only"
+end
+function get_distribution(d::Parameterized)
+    return d.distribution
+end
+
+"""
+    function sample_distribution(pds::ParameterDistributions)
+
+Draws samples from the parameter distributions
+"""
+function sample_distribution(pds::ParameterDistributions)
+    return sample_distribution(pds,1)
+end
+
+function sample_distribution(pds::ParameterDistributions, n_draws::IT) where {IT <: Integer}
+    return cat([sample_distribution(pd,n_draws) for pd in pds.parameter_distributions]...,dims=1)
+end
+
+function sample_distribution(pd::ParameterDistribution)
+    return sample_distribution(pd,1)
+end
+
+function sample_distribution(pd::ParameterDistribution, n_draws::IT) where {IT <: Integer}
+    return sample_distribution(pd.distribution,n_draws)
+end
+
+function sample_distribution(d::Samples, n_draws::IT)  where {IT <: Integer}
+    n_stored_samples = n_samples(d)
+    samples_idx = StatsBase.sample(collect(1:n_stored_samples), n_draws)
+    return d.distribution_samples[:,samples_idx]
+
+end
+
+function sample_distribution(d::Parameterized, n_draws::IT) where {IT <: Integer}
+    return rand(d.distribution, n_draws)
+end
+
+
+#apply transforms
+
```

Comment:
I find the terms real <-> prior could be confusing. For example, I may think of the prior distribution as also being in the real space, since this is the space where my prior belief is based on. What do you think about `transform_bounded_to_unbounded` and `transform_unbounded_to_bounded`?

Reply:
Yeah i mulled over this for a while actually. The terminology came from these 2 perspectives
(i) The modelling assumes we are based in the `Real` space. e.g we think of the parameters are `y` in `Real` and `F: Real -> Data` is the forward model. The Bayesian methodology we developed works only on an unbounded prior space, and so from the mathematical stance we have a single map `G: Prior -> Data`. If we use a map `G` we will need first need a transformation `T: Real -> Prior`, then the action is `F(y) = G(T(y))`. 

(ii) The other perspective is that the mathematical stance is `G: Prior -> Data`, and so we define a map `S:Prior -> Real` and then we apply the model `F`, so `G(x) = F(S(x))`. 

From perspective (i) the prior is a e.g lognormal distribution and your forward map is the model, but in order to work with it you need to work in a computational space. From (ii) the prior is a normal distribution, and the forward model maps this to the data (the black box workings don't matter), I preferred the latter because the theory is based on this `Prior` space and not on the `Real` space, and this is actually how our code works. 

So coming to the naming, I don't like mentioning `bounded` because this is only 1 of the 4 cases. But you are right the one key purpose is map to an unbounded space, i do like that; but then is it clear where your `prior` distribution is defined?

Reply:
As discuss, let's see what the others think. I think if one of the two includes `unbounded` or `unconstrained`, there should be no confusion.

Reply:
I agree the naming of `transform_real_to_prior` is a bit confusing. As long as it is explicitly documented in comments and docs it could be fine, since on the surface level a user would probably assume the distribution they are providing is the one in the `Prior` space rather than the `Real` space.